### PR TITLE
refactor recommend cli

### PR DESCRIPTION
**Pull request by Prateeknandle**
items:

- using dependency injection to create a plugin system for policy generators(kubearmor, kyverno)
- refractoring recommend cli
- removing kyverno policy generator

### CODE REVIEW

**Comment by Prateeknandle** [comment]
test failing are due to new release of policy template

File: hacks/common.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package hacks
```
Comment (achrefbensaad) [inline hacks/common.go:1]: Please handle lint comments

File: recommend/engines/generic_policies/policy.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package genericpolicies
```
Comment (achrefbensaad) [inline recommend/engines/generic_policies/policy.go:1]: delete this file

**Review by achrefbensaad — APPROVED** [review]
LGTM!

**Comment by PrimalPimmy** [comment]
LGTM

File: cmd/recommend.go [hunk 3]
```diff
@@ -45,10 +45,9 @@ func init() {
...
-	recommendCmd.Flags().StringSliceVarP(&recommendOptions.Policy, "policy", "p", recommend.DefaultPoliciesToBeRecommended, "Types of policy that can be recommended: KubeArmorPolicy|KyvernoPolicy (comma separated)")
 	recommendCmd.Flags().StringVarP(&recommendOptions.Namespace, "namespace", "n", "", "User defined namespace value for policies")
 	recommendCmd.Flags().StringVarP(&recommendOptions.OutDir, "outdir", "o", "out", "output folder to write policies")
 	recommendCmd.Flags().StringVarP(&recommendOptions.ReportFile, "report", "r", "report.txt", "report file")
 	recommendCmd.Flags().StringSliceVarP(&recommendOptions.Tags, "tag", "t", []string{}, "tags (comma-separated) to apply. Eg. PCI-DSS, MITRE")
-	recommendCmd.Flags().StringVarP(&recommendOptions.Config, "config", "c", recommend.UserHome()+"/.docker/config.json", "absolute path to image registry configuration file")
+	recommendCmd.Flags().StringVarP(&recommendOptions.Config, "config", "c", genericpolicies.UserHome()+"/.docker/config.json", "absolute path to image registry configuration file")
```
Comment (daemon1024) [inline cmd/recommend.go:52]:
```suggestion
	recommendCmd.Flags().StringVarP(&recommendOptions.Config, "config", "c", common.UserHome()+"/.docker/config.json", "absolute path to image registry configuration file")
```

I think the flag should be part of common pkg not engine specific package?

File: go.mod [hunk 4]
```diff
@@ -209,6 +206,7 @@ require (
 	github.com/klauspost/pgzip v1.2.5 // indirect
 	github.com/kr/pretty v0.3.1 // indirect
 	github.com/kr/text v0.2.0 // indirect
+	github.com/kyverno/kyverno v1.9.2 // indirect
```
Comment (daemon1024) [inline go.mod:209]: why do we still have kyverno dependency?

**Review by daemon1024 — CHANGES_REQUESTED** [review]
Some comments based on a quick review.

File: go.mod (hunk 4 above)
Comment (Prateeknandle) [reply go.mod:209]: this is presnt as an indirect dependency due to `discovery-engine` package, which uses `kyverno` pkg

Comment (daemon1024) [reply go.mod:209]: We will need to wait for to remove summary to eliminate it ig. Thanks

**Comment by Prateeknandle** [comment]
FYI : policies generated also contains empty fields, the structs have `omitempty` mentioned but still empty fields are showing.
```
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  creationTimestamp: null
  name: ubuntu-latest-maint-tools-access
spec:
  action: Audit
  capabilities:
    matchCapabilities: null
  file: {}
  message: restricted maintenance tool access attempt detected
  network:
    matchProtocols: null
  process:
    matchDirectories:
    - dir: /sbin/
      recursive: true
  selector:
    matchLabels:
      kubearmor.io/container.name: ubuntu
  severity: 1
  syscalls: {}
  tags:
  - PCI_DSS
  - MITRE
status: {}

```

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: recommend/engines/generic_policies/policy-templates.go [hunk 5]
```diff
@@ -1,56 +1,43 @@
...
-// Copyright 2022 Authors of KubeArmor
-
-package recommend
+package genericpolicies
 
 import (
 	"archive/zip"
 	"context"
+	_ "embed" // need for embedding
+	"encoding/json"
```
[the remark this answers is not in the document]
Comment (Prateeknandle) [reply recommend/engines/generic_policies/policy-templates.go]: thanks

**Review by daemon1024 — APPROVED** [review]

**Pull request merged**